Add ZenFSMetrics to ZenFS

sagitrs opened the pull request:
ZenFSMetrics is used to monitor some of the variables on ZenFS in real time.
In addition to this, two other classes have been added to metrics.h.
NoZenFSMetrics is an empty implementation of ZenFSMetrics and is used by default.
ZenFSMetricsLatencyGuard is used to count the execution time of a function, and it reports the time between construction and destruction to Metrics.

metaspace commented:
@sagitrs Tank you for the PR! Will review it this week if possible.

skyzh commented on fs/zbd_stat.h:
```diff
@@ -0,0 +1,33 @@
+// Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved.
```
Is this file expected to be added in this PR?

sagitrs replied:
Sorry, this file will be used for another PR.

skyzh commented on fs/fs_zenfs.cc:
```diff
@@ -1081,6 +1088,9 @@ Status NewZenFS(FileSystem** fs, const std::string& bdevname) {
   }
+  // This line is not necessary if you do not plan to use Metrics,
+  // as NoZenFSMetrics will be used by default.
+  //zbd->SetMetrics(std::make_shared<NoZenFSMetrics>());
```
So adding `zbd->SetMetrics(std::make_shared<NoZenFSMetrics>());`doesn't make any sense? I think this comment is not necessary.

skyzh commented on fs/metrics.h:
```diff
@@ -0,0 +1,106 @@
+#pragma once
```
New `.h` file should be added to https://github.com/westerndigitalcorporation/zenfs/blob/master/zenfs.mk

sagitrs replied:
OK.

skyzh commented on fs/metrics.h:
```diff
@@ -0,0 +1,106 @@
+public:
+  // Add a reporter named label.
+  // You can give a type for type-checking.
+  virtual void AddReporter(uint32_t label, uint32_t type = 0) = 0;
```
What does type-checking mean?

sagitrs replied:
Different Reporter may have different types, and even if the values are of the same type, the units may be different. For example, we may have created a Reporter with QPS but called it with throughput X passed in. to prevent errors caused by this situation, we plan not to use Report() directly, but use ReportXXXX() with type-checking instead.

skyzh replied:
Should the `type` be a enum? e.g. `TYPE_TIME_MS`, `TYPE_QPS`, etc.

sagitrs replied:
Yes, it should.

skyzh replied:
Just found `ZenFSMetricsReporterType`. Better to document it in the comments.

royguo commented:
@yhr Please take a look, I believe this patch doesn't hurt the performance and could help us do better metrics reporting. Users like TerarkDB or RocksDB could implement their own ZenFSMetrics.

skyzh commented on fs/metrics.h:
```diff
@@ -0,0 +1,106 @@
+  ZENFS_FG_SYNC_LATENCY,
+  ZENFS_BG_SYNC_LATENCY,
+  ZENFS_IO_ALLOC_WAL_LATENCY,
+  ZENFS_IO_ALLOC_NON_WAL_LATENCY,
```
It seems that `WAL_LATENCY` and `NON_WAL_LATENCY` is not useful in this ZenFS. These are distinguished from each other only in TerarkDB's own fork of ZenFS.

royguo replied:
> 

@skyzh Noticed, thanks. I think we could make use of these flags soon. We are now moving from our own fork to the official ZenFS, so more patches are coming (don't worry about the waste of these flags, for now, LOL).

skyzh reviewed (commented):
Rest LGTM. Did you run `clang-format` for this patch?

yhr commented:
@royguo  Had a quick look and i think the solution looks great, very neat, we might as well get this into master (as the change  does not look breaking in any way)

@sagitrs : Could you
- squash the patches into one (using git rebase -i for example)  with a subject like "fs: Add ZenFSMetrics to ZenFS"
- add some commit text that explains what you changed and why
- add your signoff (git  commit --amend --signoff)
- submit this PR to the master branch

Thanks!

sagitrs commented:
> 



> @royguo Had a quick look and i think the solution looks great, very neat, we might as well get this into master (as the change does not look breaking in any way)
> 
> @sagitrs : Could you
> 
> * squash the patches into one (using git rebase -i for example)  with a subject like "fs: Add ZenFSMetrics to ZenFS"
> * add some commit text that explains what you changed and why
> * add your signoff (git  commit --amend --signoff)
> * submit this PR to the master branch
> 
> Thanks!

I'll do it soon :)

sagitrs commented:
> @royguo Had a quick look and i think the solution looks great, very neat, we might as well get this into master (as the change does not look breaking in any way)
> 
> @sagitrs : Could you
> 
> * squash the patches into one (using git rebase -i for example)  with a subject like "fs: Add ZenFSMetrics to ZenFS"
> * add some commit text that explains what you changed and why
> * add your signoff (git  commit --amend --signoff)
> * submit this PR to the master branch
> 
> Thanks!

I think I completed the process, is that right?

> 

I think I completed the process, is that right?

skyzh commented:
> submit this PR to the master branch

Looks like the base branch is still `for_v2`.

sagitrs commented:
> @royguo Had a quick look and i think the solution looks great, very neat, we might as well get this into master (as the change does not look breaking in any way)
> 
> @sagitrs : Could you
> 
> * squash the patches into one (using git rebase -i for example)  with a subject like "fs: Add ZenFSMetrics to ZenFS"
> * add some commit text that explains what you changed and why
> * add your signoff (git  commit --amend --signoff)
> * submit this PR to the master branch
> 
> Thanks!



> @royguo Had a quick look and i think the solution looks great, very neat, we might as well get this into master (as the change does not look breaking in any way)
> 
> @sagitrs : Could you
> 
> * squash the patches into one (using git rebase -i for example)  with a subject like "fs: Add ZenFSMetrics to ZenFS"
> * add some commit text that explains what you changed and why
> * add your signoff (git  commit --amend --signoff)
> * submit this PR to the master branch
> 
> Thanks!



> > submit this PR to the master branch
> 
> Looks like the base branch is still `for_v2`.

I changed the base of PR from for_v2 to master, is that right?

skyzh commented:
>  changed the base of PR from for_v2 to master, is that right?

Yes, but now your PR includes two commits. You should rebase and remove that commit from v2.

```
git rebase -i master
```

Then use the editor to remove the extra commit.

metaspace commented on fs/fs_zenfs.cc:
```diff
@@ -232,7 +232,7 @@ void ZenFS::LogFiles() {
     for (unsigned int i = 0; i < extents.size(); i++) {
       ZoneExtent* extent = extents[i];
-      Info(logger_, "          Extent %u {start=0x%lx, zone=%u, len=%u} ", i,
+      Info(logger_, "          Extent %u {start=0x%lx, zone=%u, len=%lu} ", i,
```
@sagitrs I think you included this commit by accident. It should stay in the `for_v2` branch :)

metaspace commented on fs/metrics.h:
```diff
@@ -0,0 +1,111 @@
+  // and more
+};
+
+struct NoZenFSMetrics : public ZenFSMetrics {
```
A default implementation (in addition to `NoZenFSMetrics`) that actually records and reports metrics would be great. It would make the code a lot more useful, and would be a nice example of how to implement custom metric collection.

sagitrs replied:
> 

I've given a default implementation as a sample, so hopefully this will help.

metaspace replied:
Great, thanks! I think we could eventually make a config item to enable the default reporter.

metaspace commented on fs/fs_zenfs.cc:
```diff
@@ -1081,6 +1088,8 @@ Status NewZenFS(FileSystem** fs, const std::string& bdevname) {
     return Status::IOError(zbd_status.ToString());
   }
+  
+  zbd->SetMetrics(std::make_shared<NoZenFSMetrics>());
```
Metrics are already set to `NoZenFSMetrics` during construction of `zbd`.

sagitrs replied:
Yes, this statement is more of a usage suggestion, i.e. "where you should stuff your own implementation of Metrics". Would it be better to remove it?

metaspace replied:
I think so.

metaspace commented on fs/zbd_zenfs.cc:
```diff
@@ -174,7 +174,8 @@ Zone *ZonedBlockDevice::GetIOZone(uint64_t offset) {
                                    std::shared_ptr<Logger> logger)
-    : filename_("/dev/" + bdevname), logger_(logger) {
+    : filename_("/dev/" + bdevname), logger_(logger),
+      metrics_(std::make_shared<NoZenFSMetrics>()) {
```
Could we optionally inject a `ZenFSMetrics` instance here? But default to `NoZenFSMetrics`. I don't think it makes sense to change the metrics instance after zbd construction, so we could drop `ZonedBlockDevice::SetMetrics()`.

sagitrs replied:
Yes, the user should provide Metrics at initialization time, not SetMetrics at an unknown time. 
I believe it would be much better to add a constructor that can initialize Metrics.

sagitrs replied:
I added an initialization function for ZoneBlockDevice to pass in Metrics, and a version for NewZenFS with a Metrics parameter.

metaspace replied:
I don't think we need a new constructor. We can just provide a default argument for the existing constructor:

```suggestion
ZonedBlockDevice::ZonedBlockDevice(std::string bdevname,
                                   std::shared_ptr<Logger> logger,
                                   std::shared_ptr<ZenFSMetrics> metrics = std::make_shared<NoZenFSMetrics>())
    : filename_("/dev/" + bdevname), logger_(logger), metrics_(metrics) {
...
}
```

metaspace commented on fs/zbd_zenfs.h:
```diff
@@ -127,6 +131,9 @@ class ZonedBlockDevice {
 
   std::mutex zone_resources_mtx_; /* Protects active/open io zones */
 
+  void SetMetrics(std::shared_ptr<ZenFSMetrics> metrics) { metrics_ = metrics; }
```
I think we should drop this method and inject metric collection at construction time.

metaspace commented on fs/fs_zenfs.cc:
```diff
@@ -281,6 +281,10 @@ IOStatus ZenFS::RollMetaZoneLocked() {
   Zone *new_meta_zone, *old_meta_zone;
   IOStatus s;
 
+  ZenFSMetricsLatencyGuard guard(zbd_->GetMetrics(), 
```
Please run the linter:
```shell
clang-format --style=file -i fs_zenfs.cc io_zenfs.cc zbd_zenfs.cc fs_zenfs.h io_zenfs.h metrics.h zbd_zenfs.h
```

sagitrs replied:
Done.

metaspace reviewed (changes requested):
Thanks for the PR, I think it is a good addition.  Please see my inline comments.

sagitrs commented:
> 
I see, that means I should make my commit "look like" it was forked from master, not for_v2.
Wouldn't using "git rebase --onto master for_v2 add-metrics-patch" give the same result?

skyzh commented:
> I see, that means I should make my commit "look like" it was forked from master, not for_v2.
> Wouldn't using "git rebase --onto master for_v2 add-metrics-patch" give the same result?

`git rebase -i` is my preferred way. You can use any method that retains your only commit in your branch as you like 🥳

skyzh commented on fs/io_zenfs.h:
```diff
@@ -97,9 +97,13 @@ class ZoneFile {
   uint64_t GetID() { return file_id_; }
   size_t GetUniqueId(char* id, size_t max_size);
 
+<<<<<<< HEAD
```
Please resolve conflicts.

sagitrs replied:
Finished.

metaspace commented:
> I see, that means I should make my commit "look like" it was forked from master, not for_v2. Wouldn't using "git rebase --onto master for_v2 add-metrics-patch" give the same result?

@sagitrs  one way to do it is like so:
```shell
git remote add upstream https://github.com/westerndigitalcorporation/zenfs/
git checkout -b upstream-master upstream/master
git cherry-pick ffe0aa965559b81a98ca16f5c384f0450b98c1f6
```

If the patch does not apply, you will need to resolve any conflicts. If you prefer, I can also do the rebase for you.

metaspace commented on fs/metrics_sample.h:
```diff
@@ -0,0 +1,208 @@
+                           uint32_t type_uint = 0) override {
+    auto label = static_cast<ZenFSMetricsHistograms>(label_uint);
+    assert(ZenFSHistogramsNameMap.find(label) !=
+           BDZenFSHistogramsNameMap.end());
```
This looks like a typo

sagitrs replied:
yes.
I may have compiled it without saving it, so I didn't catch the error.

yhr reviewed (changes requested):
You need to add license headers to the new files

See this example for the preferred way of doing this for new files, using spdx-license identifiers:
https://github.com/westerndigitalcorporation/zenfs/blob/master/util/zenfs.8

yhr commented:
One small comment from me, so if this looks ok to you @metaspace , lets pull it in once the license text is added.

metaspace commented on fs/zbd_zenfs.h:
```diff
@@ -94,12 +95,15 @@ class ZonedBlockDevice {
   explicit ZonedBlockDevice(std::string bdevname,
-                            std::shared_ptr<Logger> logger);
+                            std::shared_ptr<Logger> logger,
+                            std::shared_ptr<ZenFSMetrics> metrics = nullptr);
```
Not sure, but I think that when you deref a shared_ptr that holds a nullptr, you get a segfault. We should probably initialize to `std::make_shared<NoZenFSMetrics>()` instead.

sagitrs replied:
Yes, this is a serious error. Modified.

metaspace commented on fs/fs_zenfs.h:
```diff
@@ -353,6 +356,8 @@ class ZenFS : public FileSystemWrapper {
 
 Status NewZenFS(FileSystem** fs, const std::string& bdevname);
+Status NewZenFS(FileSystem** fs, const std::string& bdevname,
+                std::shared_ptr<ZenFSMetrics> metrics);
```
I think we should use just one constructor and then a default argument to `NoZenFSMetrics`, just as for `ZonedBlockDevice`

sagitrs replied:
Done.

metaspace reviewed (changes requested):
Thank you for updating the PR. I have just a few more comments about the constructors that inject the metrics instance. I think it would be great if we just had one constructor with a default argument.

sagitrs commented:
> You need to add license headers to the new files
> 
> See this example for the preferred way of doing this for new files, using spdx-license identifiers: https://github.com/westerndigitalcorporation/zenfs/blob/master/util/zenfs.8

Hopefully I am done (as this is the first time I have added license headers to files).
I added the spdx-license identifier for metrics_sample.h in the given format, and added the same license header for metrics.h as for the other files.

yhr commented:
Merged after rebase as https://github.com/westerndigitalcorporation/zenfs/commit/7cad1d7cfa123a739f0d3364f9042120cc9985de

yhr commented:
Thanks!
